File: programmers/stack_queue/주식가격.cpp
#include <vector>
using namespace std;

vector<int> solution(vector<int> prices) {
    int size = prices.size();
    vector<int> answer(size);
    vector<pair<int,int>> stack; // 인덱스, 값
    stack.emplace_back(0, prices[0]);
  
    for (int i = 1; i < size; i++) {
        while(prices[i] < stack.back().second) {
            answer[stack.back().first] = i - stack.back().first;
            stack.pop_back();
        }
        stack.emplace_back(i, prices[i]);
    }

    while(!stack.empty()) {
        int index = stack.back().first;
        answer[index] = size - index - 1;
        stack.pop_back();
    }

    return answer;
}

File: 프로그래머스/lv2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        Stack<Integer> st = new Stack<>();
        for (int i=0; i<prices.length; i++) {
            while (!st.isEmpty() && prices[i] < prices[st.peek()]) {
                int j = st.pop();
                answer[j] = i - j;
            }
            st.push(i);
        }
        while (!st.isEmpty()) {
            int i = st.pop();
            answer[i] = prices.length - 1 - i;
        }
        
        return answer;
    }
}